Handling multiple tags for one node scenerio

yuquanzhou opened the pull request:
Refactored the code to handle multiple tags for one node scenerio and fixed minor bugs

yuquanzhou committed:
Handling multiple tags for one node scenerio

Refactored the code to handle multiple tags for one node scenerio and fixed minor bugs

gboeing commented:
@yuquanzhou please list and summarize in a comment the enhancements and fixes this PR is proposing.

yuquanzhou commented:
**1. Handled the situation where a single node has multiple highway tags**
- The previous version assumes that there is only one single highway tag per node. 
- The updated `add_edge_control_non_simplified()` function handles the situation where a single node has multiple highway tags by parsing multi-tag strings (e.g., `"traffic_signals;crossing"`) into lists (e.g. `["traffic_signals", "crossing"]`) and assigning it to edges.

**2. Created new `control_applies()` function**
- Although the previous script has already considered the directions of the traffic control would apply, but this added function makes the code more succinct without repeating it every time to judge the direction of each kind of traffic control.
- It judges whether the traffic control applies to edge based on the traffic control's direction and edge direction.

**3. Removed `turning_circle` tag**
-  The `turning_circle` in the highway tag refers to the turning circle at the roads' dead-ends.
-  We didn't include it in our model because it is not impeding the traffic. Therefore, I also deleted in the script.

**4. More safeguard measures to prevent running failure due to invalid input**
-  Valid OD node filtering before calculate the routing: only calculate the routing for OD pairs with OID and DID in the graph.
-  Try-except block wrapping routing logic (`except (ValueError, NodeNotFound, NetworkXNoPath, KeyError)`):  catches expected routing-related exceptions and safely returns a dummy result.

**5. Minor bugs fix**
- Duplicate turn names `_RIGHT_TURN_DEGREE_LOWER_BOUND` for slight turn
- Differentiate that we used simplified graph for OD pair sampling and unsimplified graph for routing in `01-preprocess_input_data.py`.
